### PR TITLE
Cramers v

### LINKED ISSUE #1272: [FEAT] Add Cramer's V in addition to PSI for all categorical drift

### PR DESCRIPTION
**Pull request by JKL98ISR**
closes #1272

### CODE REVIEW

File: deepchecks/tabular/checks/distribution/train_test_feature_drift.py [hunk 1]
```diff
@@ -218,10 +229,10 @@ def add_condition_drift_score_not_greater_than(self, max_allowed_psi_score: floa
...
         Parameters
         ----------
-        max_allowed_psi_score: float , default: 0.2
-            the max threshold for the PSI score
-        max_allowed_earth_movers_score: float , default: 0.1
-            the max threshold for the Earth Mover's Distance score
+        max_allowed_categorical_score: float , default: 0.2
+            the max threshold for the categorical drift score
+        max_allowed_numerical_score: float ,  default: 0.1
+            the max threshold for the numerical drift score
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/distribution/train_test_feature_drift.py:236]: it's not "numerical drift" and "categorical drift". it's "numeric variable drift score" and "categorical variable drift score"

File: deepchecks/tabular/checks/distribution/train_test_feature_drift.py [hunk 2]
```diff
@@ -242,24 +253,26 @@ def condition(result: Dict) -> ConditionResult:
...
+                                           and d['Method'] in SUPPORTED_NUMERICAL_METHODS
                                            and column in columns_to_consider}
             return_str = ''
             if not_passing_categorical_columns:
-                return_str += f'Found categorical columns with PSI above threshold: {not_passing_categorical_columns}\n'
+                return_str += 'Found categorical columns with drift score above threshold: ' \
+                              f'{not_passing_categorical_columns}\n'
             if not_passing_numeric_columns:
-                return_str += f'Found numeric columns with Earth Mover\'s Distance above threshold: ' \
+                return_str += f'Found numerical columns with drift score above threshold: ' \
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/distribution/train_test_feature_drift.py:269]:
```suggestion
                return_str += f'Found numeric columns with drift score above threshold: ' \
```

File: deepchecks/tabular/checks/distribution/train_test_feature_drift.py [hunk 3]
```diff
@@ -242,24 +253,26 @@ def condition(result: Dict) -> ConditionResult:
...
                               f'{not_passing_numeric_columns}'
 
             if return_str:
                 return ConditionResult(ConditionCategory.FAIL, return_str)
             else:
                 return ConditionResult(ConditionCategory.PASS)
 
-        return self.add_condition(f'PSI <= {max_allowed_psi_score} and Earth Mover\'s Distance <= '
-                                  f'{max_allowed_earth_movers_score}',
+        return self.add_condition(f'categorical drift score <= {max_allowed_categorical_score} and '
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/distribution/train_test_feature_drift.py:277]: Use the method's name. In general, even if I miss this in this PR, do not use "categorical/numeric drift score".

File: deepchecks/tabular/checks/distribution/train_test_label_drift.py [hunk 4]
```diff
@@ -49,6 +55,8 @@ class TrainTestLabelDrift(TrainTestCheck):
         - 'train_largest': Show the largest train categories.
         - 'test_largest': Show the largest test categories.
         - 'largest_difference': Show the largest difference between categories.
+    categorical_drift_method: str, default: "Cramer"
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/distribution/train_test_label_drift.py:58]: Consulted with Bressler, let's change to "cramers_v"?

File: deepchecks/tabular/checks/distribution/train_test_label_drift.py [hunk 5]
```diff
@@ -49,6 +55,8 @@ class TrainTestLabelDrift(TrainTestCheck):
...
+        Cramer for Cramer's V, PSI for Population Stability Index (PSI).
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/distribution/train_test_label_drift.py:59]: Should first describe what this does. "decides which method to use on categorical variables. Possible values are: "

File: deepchecks/utils/distribution/drift.py [hunk 6]
```diff
@@ -24,9 +24,45 @@
...
+    contingency_matrix = pd.DataFrame([dist1_counts, dist2_counts])
+
+    chi2 = chi2_contingency(contingency_matrix)[0]
+    n = contingency_matrix.sum().sum()
+    phi2 = chi2/n
+    r, k = contingency_matrix.shape
+    phi2corr = max(0, phi2 - ((k-1)*(r-1))/(n-1))
+    rcorr = r - ((r-1)**2)/(n-1)
+    kcorr = k - ((k-1)**2)/(n-1)
+    return np.sqrt(phi2corr / min((kcorr-1), (rcorr-1)))
```
Comment (nirhutnik) [inline deepchecks/utils/distribution/drift.py:65]: This is based on https://stackoverflow.com/questions/46498455/categorical-features-correlation/46498792#46498792  and is reused in other sources ( https://towardsdatascience.com/the-search-for-categorical-correlation-a1cf7f1888c9 ) and I think we need to give credit. @noamzbr ?

File: deepchecks/utils/distribution/drift.py [hunk 7]
```diff
@@ -177,18 +216,32 @@ def calc_drift_and_plot(train_column: pd.Series,
...
-        )
+        if categorical_drift_method == 'Cramer':
+            scorer_name = 'Cramer\'s V'
+            score = cramers_v(dist1=train_dist, dist2=test_dist)
+
+            bar_traces, bar_x_axis, bar_y_axis = drift_score_bar_traces(score, bar_max=1)
+            dist_traces, dist_x_axis, dist_y_axis = feature_distribution_traces(
+                train_dist, test_dist, value_name, is_categorical=True,
+                max_num_categories=max_num_categories_for_display,
+            )
```
Comment (nirhutnik) [inline deepchecks/utils/distribution/drift.py:227]: the last 5 lines of code could be reused between both methods...

File: deepchecks/vision/checks/distribution/train_test_label_drift.py [hunk 8]
```diff
@@ -65,7 +70,7 @@ class TrainTestLabelDrift(TrainTestCheck):
         out of the EMD calculation. This is done in order for extreme values not to affect the calculation
         disproportionally. This filter is applied to both distributions, in both margins.
     max_num_categories_for_drift: int, default: 10
-        Only for non-continuous properties. Max number of allowed categories. If there are more,
+        Only for categorical properties. Max number of allowed categories. If there are more,
```
Comment (nirhutnik) [inline deepchecks/vision/checks/distribution/train_test_label_drift.py:73]: For now, write "discrete". In the next task where we rewrite the "output_type" thingy then we'll change this docstring as well

File: tests/vision/checks/distribution/image_dataset_drift_test.py [hunk 9]
```diff
@@ -30,7 +30,29 @@ def pil_drift_formatter(batch):
...
+            'Mean Blue Relative Intensity': equal_to(0),
+            'Mean Green Relative Intensity': equal_to(0),
+        })
+    }))
+
+
+def test_no_drift_grayscale_cramer(mnist_dataset_train, device):
+    # Arrange
+    train, test = mnist_dataset_train, mnist_dataset_train
+    check = ImageDatasetDrift(categorical_drift_method='Cramer')
```
Comment (nirhutnik) [inline tests/vision/checks/distribution/image_dataset_drift_test.py:55]: ummmmm what?
This check does not use PSI or Cramer's V or EMD... ImageDatasetDrift and WholeDatasetDrift work in completely different ways... This should have no effect.

File: tests/vision/checks/distribution/image_property_drift_test.py [hunk 10]
```diff
@@ -37,7 +37,22 @@ def test_image_property_drift_check(coco_train_visiondata, coco_test_visiondata,
...
+    result = ImagePropertyDrift(categorical_drift_method='PSI',
+                                classes_to_display=['bicycle', 'bench', 'bus', 'truck'], min_samples=5
+                                ).run(coco_train_visiondata, coco_test_visiondata, device=device)
+
+    # Assert
+    assert_that(result, is_correct_image_property_drift_result())
+
+    assert_that(result.value, has_entries(
+        {'Brightness': close_to(0.13, 0.01)}
+    ))
```
Comment (nirhutnik) [inline tests/vision/checks/distribution/image_property_drift_test.py:49]: Currently, all image properties are continuous / numeric. The categorical_drift_method parameter should have no effect

File: tests/vision/checks/distribution/train_test_prediction_drift_test.py [hunk 11]
```diff
@@ -166,13 +185,13 @@ def infer(batch, model, device):
     result = check.run(mod_train_ds, mod_test_ds, mock_trained_mnist, device=device, n_samples=None)
 
     condition_result, *_ = result.conditions_results
-
+    print(condition_result)
```
Comment (nirhutnik) [inline tests/vision/checks/distribution/train_test_prediction_drift_test.py:188]: whoopsy

File: deepchecks/utils/distribution/drift.py (hunk 7 above)
Comment (JKL98ISR) [reply deepchecks/utils/distribution/drift.py:227]: Psi is a bit different, i can refactor the pso function tho

File: deepchecks/tabular/checks/distribution/train_test_feature_drift.py (hunk 3 above)
Comment (JKL98ISR) [reply deepchecks/tabular/checks/distribution/train_test_feature_drift.py:277]: I don't know which methods will be used in the check

File: deepchecks/tabular/checks/distribution/train_test_label_drift.py (hunk 4 above)
Comment (JKL98ISR) [reply deepchecks/tabular/checks/distribution/train_test_label_drift.py:58]: earth_movers has naming with _

File: deepchecks/utils/distribution/drift.py (hunk 6 above)
Comment (JKL98ISR) [reply deepchecks/utils/distribution/drift.py:68]: added for now

File: deepchecks/tabular/checks/distribution/train_test_feature_drift.py (hunk 3 above)
Comment (nirhutnik) [reply deepchecks/tabular/checks/distribution/train_test_feature_drift.py:278]: Okay, right. So let's keep it like that, but make sure that in the return_str of the condition it mentions the methods

File: deepchecks/tabular/checks/distribution/train_test_label_drift.py (hunk 4 above)
Comment (nirhutnik) [reply deepchecks/tabular/checks/distribution/train_test_label_drift.py:58]: What?

File: deepchecks/tabular/checks/distribution/train_test_feature_drift.py [hunk 12]
```diff
@@ -71,6 +71,7 @@ class TrainTestFeatureDrift(TrainTestCheck):
         - 'test_largest': Show the largest test categories.
         - 'largest_difference': Show the largest difference between categories.
     categorical_drift_method: str, default: "Cramer"
+        decides which method to use on categorical variables. Possible values are:
         Cramer for Cramer's V, PSI for Population Stability Index (PSI).
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/distribution/train_test_feature_drift.py:75]: "cramers_v" for ... , "PSI" for ...

**Review by nirhutnik — APPROVED** [review]

File: deepchecks/tabular/checks/distribution/train_test_feature_drift.py [hunk 13]
```diff
@@ -107,6 +117,7 @@ def __init__(
         else:
             raise DeepchecksValueError('sort_feature_by must be either "feature importance" or "drift score"')
         self.n_top_columns = n_top_columns
+        self.categorical_drift_method = categorical_drift_method
```
Comment (matanper) [inline deepchecks/tabular/checks/distribution/train_test_feature_drift.py:121]: maybe validate the value in the init?

File: deepchecks/utils/distribution/drift.py [hunk 14]
```diff
@@ -24,9 +24,79 @@
...
+    Tuple(str, str)
+        the categorical scoring method and then the numeric scoring method.
+
+    """
+    result_df = pd.DataFrame(result_dict).T
+    cat_mthod_arr = result_df[result_df['Method'].isin(SUPPORTED_CATEGORICAL_METHODS)]['Method']
+    if len(cat_mthod_arr):
+        cat_method = cat_mthod_arr[0]
+    else:
+        cat_method = None
```
Comment (matanper) [inline deepchecks/utils/distribution/drift.py:53]: possible to reduce to one-liner
```suggestion
   cat_method = cat_mthod_arr[0] if cat_mthod_arr else None
```

File: deepchecks/utils/distribution/drift.py [hunk 15]
```diff
@@ -24,9 +24,79 @@
...
+
+    num_mthod_arr = result_df[result_df['Method'].isin(SUPPORTED_NUMERIC_METHODS)]['Method']
+    if len(num_mthod_arr):
+        num_method = num_mthod_arr[0]
+    else:
+        num_method = None
```
Comment (matanper) [inline deepchecks/utils/distribution/drift.py:59]: same as above

**Review by matanper — APPROVED** [review]

File: deepchecks/tabular/checks/distribution/train_test_feature_drift.py (hunk 13 above)
Comment (JKL98ISR) [reply deepchecks/tabular/checks/distribution/train_test_feature_drift.py:121]: in similar cases we also don't check on init so I don't think this case should be an exception

**Review by ItayGabbay — APPROVED** [review]

**Comment by ItayGabbay** [comment]
Approved the docs part.

**Pull request merged**